src/backends/mod.rs

```diff
@@ -35,10 +35,205 @@
 pub mod perf_pt;
 #[cfg(perf_pt)]
 pub use self::perf_pt::PerfPTTracer;
 
+#[derive(Debug, PartialEq)]
+pub enum BackendKind {
+    Dummy,
+    PerfPT,
+}
+
+/// Generic configuration interface for all backends.
+/// A `None` field indicates a default value and attributes which don't apply to a given backend
+/// are checked by the `TracerBuilder` when `build()` is called.
+#[derive(Debug)]
+pub struct BackendConfig {
+    perf_aux_bufsize: Option<usize>,
+    perf_data_bufsize: Option<usize>,
+    new_trace_bufsize: Option<usize>,
+}
+
+/// A builder interface for instantiating backends.
+pub struct TracerBuilder {
+    user_backend: Option<BackendKind>,  // User-requested backend choice.
+    auto_backend: BackendKind,          // Auto-chosen backend if `user_backend` is `None`.
+    config: BackendConfig,
+}
+
+impl TracerBuilder {
+    /// Create a new TracerBuilder with default options.
+    pub fn new() -> Self {
+        Self {
+            user_backend: None,
+            auto_backend: Self::default_platform_backend(),
+            config: BackendConfig {
+                perf_aux_bufsize: None,
+                perf_data_bufsize: None,
+                new_trace_bufsize: None,
+            }
+        }
+    }
+
+    /// Overide auto-default backend selection.
+    pub fn backend(mut self, backend: BackendKind) -> Self {
+        self.user_backend = Some(backend);
+        self
+    }
+
+    /// Set the Linux Perf data buffer size (in pages).
+    /// Applies to only backends using Linux perf.
+    pub fn perf_data_bufsize(mut self, size: usize) -> Self {
+        self.config.perf_data_bufsize = Some(size);
+        self
+    }
+
+    /// Set the Linux Perf AUX buffer size (in pages).
+    /// Applies to only backends using Linux perf.
+    pub fn perf_aux_bufsize(mut self, size: usize) -> Self {
+        self.config.perf_aux_bufsize = Some(size);
+        self
+    }
+
+    /// Set the initial trace storage buffer size (in bytes).
+    /// Applies to all backends.
+    pub fn new_trace_bufsize(mut self, size: usize) -> Self {
+        self.config.new_trace_bufsize = Some(size);
+        self
+    }
+
+    /// Query which backend will be used.
+    /// Useful if you've allowed the builder to choose an auto-default, but wish to know which.
+    pub fn backend_kind(&self) -> &BackendKind {
+        if let Some(ref backend) = self.user_backend {
+            &backend
+        } else {
+            &self.auto_backend
+        }
+    }
+
+    // Finds a suitable `BackendKind` for the current hardware/OS.
+    fn default_platform_backend() -> BackendKind {
+        let tr_kinds = vec![BackendKind::PerfPT, BackendKind::Dummy];
+        for kind in tr_kinds {
+            if Self::match_platform(&kind).is_ok() {
+                return kind;
+            }
+        }
+        // The Dummy backend should always be usable.
+        unreachable!();
+    }
+
+    /// Returns `Ok` if the tracer backend should work on the current hardware/OS.
```

Comment:
This is a weird comment given the function name. Should the function be renamed?

Reply:
`matches_platform` would make more sense?